selfhost/binding.wyv

```diff
@@ -20,25 +21,25 @@ def bindList(es:List[raw.Exp], ctx:bound.Context):List[bound.Exp] = match es:
 def bind(e:raw.Exp, parse: String -> raw.Exp):bound.Exp
     bindHelper(e, bound.emptyContext(parse))
 
-def bindDeclTypes(e:raw.Exp, context:bound.Context):List[bound.DeclType] = match e:
-    s:raw.Seq     => llist.flatten[bound.DeclType](s.exps.map[List[bound.DeclType]](x => bindDeclTypes(x, context)))
+def bindDeclTypes(e:raw.Exp, context:bound.Context):List[types.DeclType] = match e:
+    s:raw.Seq     => llist.flatten[types.DeclType](s.exps.map[List[types.DeclType]](x => bindDeclTypes(x, context)))
     v:raw.ValType =>
-        val b = bound.Binding(v.name, context.counter)
-        llist.Singleton[bound.DeclType](bound.ValType(b,v.typ))
+        val b = types.Binding(v.name, context.counter)
+        llist.Singleton[types.DeclType](types.ValType(b,types.stringToType(v.typ)))
     default       => error.report("unexpected construct", error.unknownLocation)
     
 def bindHelper(e:raw.Exp, context:bound.Context):bound.Exp = match e:
-    v:raw.Var    => bound.Var(llist.find[bound.Binding](context.bindings, (b:bound.Binding) => v.name==b.name).get())
+    v:raw.Var    => bound.Var(llist.find[types.Binding](context.bindings, (b:types.Binding) => v.name==b.name).get())
     a:raw.App    => bound.App(bindHelper(a.func, context), bindHelper(a.arg, context))
     l:raw.Lambda =>
-        val b = bound.Binding(l.name, context.counter)
+        val b = types.Binding(l.name, context.counter)
         bound.Lambda(b, bindHelper(l.body, context.extend(b)))
     c:raw.Call   => bound.Call(bindHelper(c.receiver, context), c.name, bindHelper(c.arg, context))
     v:raw.Val    =>
-        val b = bound.Binding(v.name, context.counter)
+        val b = types.Binding(v.name, context.counter)
         bound.Val(b, bindHelper(v.exp, context))
     t:raw.TypeDecl   =>
-        val b = bound.Binding(t.name, context.counter)
+        val b = types.Binding(t.name, context.counter)
```

Comment:
should extend the context with b before passing it down to bindDeclTypes, since type t might be used within itself (like Int is used within Int)